hal/armv7a/zynq7000/pmap.c

```diff
@@ -0,0 +1,584 @@
+		return EOK;
+	}
+
+	if (page->addr >= ((addr_t)pmap_common.sptab - VADDR_KERNEL + min) && page->addr < ((addr_t)pmap_common.sptab - VADDR_KERNEL + min + SIZE_PAGE)) {
+		page->flags |= PAGE_FREE;
+		return EOK;
+	}
+
+	page->flags &= ~PAGE_FREE;
+
+	return EOK;
+}
+
+
+/* Function allocates page tables for kernel space */
+int _pmap_kernelSpaceExpand(pmap_t *pmap, void **start, void *end, page_t *dp)
+{
+	void *vaddr;
+
+	vaddr = (void *)((u32)(*start + SIZE_PAGE - 1) & ~(SIZE_PAGE - 1));
+	if (vaddr >= end)
+		return EOK;
+
+	if (vaddr < (void *)VADDR_KERNEL)
+		vaddr = (void *)VADDR_KERNEL;
+
+	for (; vaddr < end; vaddr += (SIZE_PAGE << 10)) {
+		if (pmap_enter(pmap, 0, vaddr, ~PGHD_PRESENT, NULL) < 0) {
+			if (pmap_enter(pmap, 0, vaddr, ~PGHD_PRESENT, dp) < 0) {
+				return -ENOMEM;
+			}
+			dp = NULL;
+		}
+		*start = vaddr;
+	}
+
+	pmap->start = (void *)VADDR_KERNEL;
+	pmap->end = end;
+
+	return EOK;
+}
+
+
+/* Function return character marker for page flags */
+char pmap_marker(page_t *p)
+{
+	if (p->flags & PAGE_FREE)
+		return '.';
+
+	return marksets[(p->flags >> 1) & 3][(p->flags >> 4) & 0xf];
+}
+
+
+int pmap_segment(unsigned int i, void **vaddr, size_t *size, int *prot, void **top)
+{
+	switch (i) {
+	case 0:
+		*vaddr = (void *)VADDR_KERNEL;
+		*size = (size_t)_etext - VADDR_KERNEL;
+		*prot = (PROT_EXEC | PROT_READ);
+		break;
+	case 1:
+		*vaddr = _etext;
+		*size = (size_t)(*top) - (size_t)_etext;
+		*prot = (PROT_WRITE | PROT_READ);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int pmap_getMapsCnt(void)
+{
+	return 0;
+}
+
+
+int pmap_getMapParameters(u8 id, void **start, void **end)
+{
+	return EOK;
+}
+
+
+void pmap_getAllocatedSegment(void *memStart, void *memStop, void **segStart, void **segStop)
+{
+	return;
+}
+
+
+/* Function initializes low-level page mapping interface */
+void _pmap_init(pmap_t *pmap, void **vstart, void **vend)
+{
+	int i;
+	void *v;
+
+	pmap_common.asidptr = 0;
+	pmap->asid_ix = 0;
+
+	for (i = 0; i < sizeof(pmap_common.asid_map) / sizeof(pmap_common.asid_map[0]); ++i) {
+		pmap_common.asid_map[i] = NULL;
+		pmap_common.asids[i] = i;
+	}
+
+	hal_spinlockCreate(&pmap_common.lock, "pmap_common.lock");
+
+	// pmap_common.minAddr = syspage->pbegin;
+	// pmap_common.maxAddr = syspage->pend;
+
+	/* Initialize kernel page table */
+	pmap->pdir = pmap_common.kpdir;
+	pmap->addr = (addr_t)pmap->pdir - VADDR_KERNEL + pmap_common.minAddr;
+
+	/* Remove initial kernel mapping */
+	for (i = 0; i < 4; ++i) {
+		pmap->pdir[(pmap_common.minAddr >> 20) + i] = 0;
+		hal_cpuInvalVA(pmap_common.minAddr + i * (1 << 2));
+	}
+
+	pmap->start = (void *)VADDR_KERNEL;
+	pmap->end = (void *)VADDR_MAX;
+
+	/* Initialize kernel heap start address */
+	(*vstart) = (void *)(((u32)_end + SIZE_PAGE - 1) & ~(SIZE_PAGE - 1));
+
+	/* First 15 pages after bss are mapped for UART1, UART2, GIC, GPT1, CCM, IOMUX, WDOG and SRC */
+	(*vstart) += 16 * SIZE_PAGE;
+	(*vend) = (*vstart) + SIZE_PAGE;
```

Comment:
Currently pmap isn't implemented for Zynq so it was copied from imx6ull without changes to make compilation successful. Taking into account that it will be changed in the future I haven't put it into a common directory.